doc/devdocs/locks.rst

```diff
@@ -0,0 +1,110 @@
+.. currentmodule:: Base
+
+****************************************************
+Proper maintenance and care of multi-threading locks
+****************************************************
+
+The following strategies are used to ensure that the code is dead-lock free:
+
+ 1. structure code such that only one lock will need to be acquired at a time
+
+ 2. always acquire shared locks in the same order, as given by the table below
+
+ 3. avoid constructs that expect to need unrestricted recursion
+
+Below are all of the locks that exist in the system and the mechanism in use
+that prevents them from being a potential deadlock.
+
+The following are definitely leaf locks (level 1), and must not try to acquire any other lock:
+
+   * safepoint
+
+       note that this lock is acquired implicitly by ``JL_LOCK`` and ``JL_UNLOCK``.
+       use the ``_NOGC`` variants to avoid that for level 1 locks.
+
+   * shared_map
+   * finalizers
+   * pagealloc
+   * flisp
+
+       flisp itself is already threadsafe, this lock only protects the ``jl_ast_context_list_t`` pool
+
+
+The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
+
+   * symbol_table
+
+The following are level 3 locks, which can only acquire level 1 or level 2 locks internally:
+
+   * dump
+   * typecache
+
+The following is a level 4 locks, which can only recurse to acquire level 1, 2, or 3 locks:
+
+   * Method->writelock
+
+     but note that this is violated by staged functions!
+
+The following are level 5 locks, which can only recurse to acquire to acquire locks at lower levels:
+
+   * MethodTable->writelock
+
+     but note that this is violated by staged functions!
+
+The following is a proposed level 6 locks, which can only recurse to acquire locks at lower levels:
+
+   * staged
+
+       this theoretical lock would create a priority inversion from the `method->writelock` (level 4),
+       but only prohibiting only running any staging function in parallel
+       (thus allowing temporary release of the MethodTable and Method locks)
+
+The following is a level 7 locks, which can only recurse to acquire locks at lower levels:
+
+   * codegen
+
+The following is an almost root locks (level end-1), meaning only the root look may be held when trying to acquire it:
+
+   * typeinf
+
+       this one is perhaps one of the most tricky ones, since type-inference can be invoked from many points
+
+The following is the root locks, meaning no other lock shall be held when trying to acquire it:
+
+   * toplevel
+
+       this should be held while attempting a top-level action (such as making a new type or defining a new method):
+       trying to obtain this lock inside a staged function will cause a deadlock condition!
```

Comment:
also should note here that it's unclear if any code can safely run in parallel with a toplevel expression